src/main/java/org/links/objecthash/ObjectHash.java

```diff
@@ -0,0 +1,318 @@
+                              + parseHex(hex.charAt(idx - 1)));
+    }
+    return h;
+  }
+
+  private static JsonType getType(Object jsonObj) {
+    if (jsonObj == JSONObject.NULL) {
+      return JsonType.NULL;
+    } else  if (jsonObj instanceof JSONArray) {
+      return JsonType.ARRAY;
+    } else if (jsonObj instanceof JSONObject) {
+      return JsonType.OBJECT;
+    } else if (jsonObj instanceof String) {
+      return JsonType.STRING;
+    } else if (jsonObj instanceof Integer || jsonObj instanceof Long) {
+      return JsonType.INT;
+    } else if (jsonObj instanceof Double) {
+      return JsonType.FLOAT;
+    } else if (jsonObj instanceof Boolean) {
+      return JsonType.BOOLEAN;
+    } else {
+      LOG.log(Level.WARNING, "jsonObj is_a " + jsonObj.getClass());
+      return JsonType.UNKNOWN;
+    }
+  }
+
+  public static ObjectHash commonJsonHash(String json)
+      throws JSONException, NoSuchAlgorithmException {
+    // TODO(phad): implement 'commonizing' of JSON values.
+    return new ObjectHash();
+  }
+
+  public static ObjectHash pythonJsonHash(String json)
+      throws JSONException, NoSuchAlgorithmException {
+    LOG.info("\n=============================================\n"
+             + "= JSON: " + json + "\n"
+             + "=============================================");
+    ObjectHash h = new ObjectHash();
+    h.hashAny(new JSONTokener(json).nextValue());
+    return h;
+  }
+
+  @Override
+  public String toString() {
+    return this.toHex();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+   if (this == other) return true;
+   if (other == null) return false;
+   if (this.getClass() != other.getClass()) return false;
+   return this.toHex().equals(((ObjectHash) other).toHex());
+  }
+
+  @Override
+  public int compareTo(ObjectHash other) {
+    return toHex().compareTo(other.toHex());
+  }
+
+  public byte[] hash() {
+    return hash;
+  }
+
+  private static String toHex(byte[] buffer) {
+    StringBuffer hexString = new StringBuffer();
+    for (int idx = 0; idx < buffer.length; ++idx) {
+      String hex = Integer.toHexString(0xff & buffer[idx]);
+      if (hex.length() == 1) {
+        hexString.append('0');
+      }
+      hexString.append(hex);
+    }
+    return hexString.toString();
+  }
+
+  public String toHex() {
+    return toHex(hash);
+  }
+
+  static String normalizeFloat(double f) {
+    // Early out for zero.
+    if (f == 0.0) {
+      return "+0:";
+    }
+    StringBuffer sb = new StringBuffer();
+    // Sign
+    sb.append(f < 0.0 ? '-' : '+');
+    if (f < 0.0) f = -f;
+    // Exponent
+    int e = 0;
+    while (f > 1) {
+      f /= 2;
+      e += 1;
+    }
+    while (f < 0.5) {
+      f *= 2;
+      e -= 1;
+    }
+    sb.append(e);
+    sb.append(':');
+    // Mantissa
+    if (f > 1 || f <= 0.5) {
+      throw new IllegalStateException("wrong range for mantissa");
+    }
+    while (f != 0) {
+      if (f >= 1) {
```

Comment:
It sure didn't terminate when I forgot to negate 'f' at first when sign is -ve ...